### PR TITLE
Decouple most of the types from GlDevice and GlResources

### PR DESCRIPTION
**Pull request by bjz**
Working towards #416.

### CODE REVIEW

File: src/gfx/lib.rs [hunk 1]
```diff
@@ -72,23 +72,23 @@ impl<D: device::Device> Graphics<D> {
...
-                      program: &ProgramHandle<GlResources>,
-                      mesh: &Mesh<GlResources>,
-                      slice: Slice<GlResources>,
+    pub fn make_batch<T: shade::ShaderParam<Resources = D::Resources>>(&mut self,
+                      program: &ProgramHandle<D::Resources>,
+                      mesh: &Mesh<D::Resources>,
+                      slice: Slice<D::Resources>,
                       state: &DrawState)
-                      -> Result<batch::RefBatch<T, GlResources>, batch::BatchError> {
+                      -> Result<batch::RefBatch<T>, batch::BatchError> {
```
Comment (kvark) [inline src/gfx/lib.rs]: Nice!

File: src/render/batch.rs [hunk 2]
```diff
@@ -78,18 +77,18 @@ pub trait Batch {
...
+    fn fill_params(&self, ::shade::ParamValues<Self::Resources>)
                    -> Result<&ProgramHandle<Self::Resources>, Self::Error>;
 }
 
-impl<'a, T: ShaderParam> Batch for (&'a mesh::Mesh<back::GlResources>, mesh::Slice<back::GlResources>,
-                                    &'a ProgramHandle<back::GlResources>, &'a T, &'a DrawState) {
-    type Resources = back::GlResources;
+impl<'a, T: ShaderParam> Batch for (&'a mesh::Mesh<T::Resources>, mesh::Slice<T::Resources>,
+                                    &'a ProgramHandle<T::Resources>, &'a T, &'a DrawState) {
+    type Resources = T::Resources;
```
Comment (kvark) [inline src/render/batch.rs]: Is this even needed?

**Comment by kvark** [comment]
@bjz I finished the review, now waiting for you to respond.

File: src/render/batch.rs (hunk 2 above)
Comment (bjz) [inline src/render/batch.rs]: We need it for the trait definition of `fill_params`.

Comment (kvark) [inline src/render/batch.rs]: Oh right.

**Pull request merged**